doc/addrwatch.8
.\" Manpage for addrwatch.
.\" Contact julius.kriukas@gmail.com to correct errors or typos.
.TH ADDRWATCH 8 "06 Jun 2012" "addrwatch 0.3" "addrwatch man page"
.SH NAME
addrwatch \- ethernet/ip address monitoring tool
.SH SYNOPSIS
.B addrwatch
.B [\-lmqv46dP?V]
.BI "[-m[" database "]]"
.BI "[-o " file ]
.BI "[-s " sqlite3.db ]
.BI "[-b " ip ]
.BI "[-H " size ]
.BI "[-r " num ]
.BI "[-p " pidfile ]
.BI "[-u " user ]
.BI "[-h " hostname ]
.BI "[--mysql-table=" table "]"
.BI "[--sqlite3-table=" table "]"
.BI "[" INTERFACE1 " " INTERFACE2 " " ... "]"
.SH DESCRIPTION
.B addrwatch
is a similar software to arpwatch. It main purpose is to monitor 
network and log ethernet/ip pairings.
.PP
To simply try out
.B addrwatch
start it without any arguments. When started without arguments it will open 
first non loopback interface and start logging events to the console without 
writing anything to disk. All discovered ethernet/ip address pairings will be
printed to
.BR stdout .
All debug, warning, and err messages will be sent to syslog 
and printed to 
.BR stderr .
.SH OPTIONS
.TP
.BI "-l, --syslog"
Log discovered ethernet/ip address pairings to syslog (daemon facility).
.TP
.BI "-o " "file" ", --output=" "file"
Write discovered ethernet/ip address pairings to a specified
.IR "file" "."
See
.B OUTPUT
section for more details.
.TP
.BI "-m[" database "], --mysql=[" database "]"
Write discovered ethernet/ip address pairings to MySQL
.IR database .
Default MySQL server, login information and database should be specified in 
.B "~/.my.cnf"
configuration file in 
.B "addrwatch"
section. Example of 
.B "~/.my.cnf"
configuration file:
.nf
.in +4
[addrwatch]
.RI "host=" "localhost"
.RI "user=" "db_user"
.RI "password=" "db_password"
.RI "database=" "db_name"
.in -4
.fi

Pairings are written in a table having such schema:
.nf
.in +4
CREATE TABLE IF NOT EXISTS `addrwatch` (
.in +4
`tstamp` timestamp NOT NULL,
`hostname` varchar(256) NOT NULL,
`interface` varchar(16) NOT NULL,
`vlan_tag` int(11) NOT NULL,
`mac_address` varchar(17) NOT NULL,
`ip_address` varchar(42) NOT NULL,
`origin` varchar(8) NOT NULL,
KEY `interface` (`interface`),
KEY `vlan_tag` (`vlan_tag`),
KEY `interface_vlan_tag` (`interface`,`vlan_tag`)
.in -4
)
.in -4
.fi
Default table name is 
.B addrwatch
unless option
.B "--mysql-table"
is specified.
.TP
.BI "--mysql-table=" "TBL"
Insert data to MySQL table 
.IR TBL .
.TP
.BI "-s " "sqlite3.db" ", --sqlite3=" "sqlite3.db"
Write discovered ethernet/ip address pairings to sqlite3 database in
.I sqlite3.db
file. Database file and tables are created during startup if does not exists.

Pairings are written in a table having such schema:
.nf
.in +4
CREATE TABLE addrwatch(
.in +4
timestamp UNSIGNED BIG INT,
interface varchar(16),
vlan_tag UNSIGNED INT,
mac_address varchar(17),
ip_address varchar(42),
origin TINYINT
.in -4
);
.in -4
.fi
.TP
.BI "-q, --quiet"
Suppress any output to 
.B stdout
and
.BR stderr .
.TP
.BI "-v, --verbose"
Enable debug messages.
.TP
.BI "-4, --ipv4-only"
Capture only IPv4 packets.
.TP
.BI "-6, --ipv6-only"
Capture only IPv6 packets.
.TP
.BI "-b " "ip" ", --blacklist=" "ip"
Ignore pairings with specified
.I ip
address.
.TP
.BI "-r " "num" ", --ratelimit=" "num"
If
.I num
> 0, ratelimit duplicate ethernet/ip pairings to 1 every
.I num
seconds.
.br
If
.I num
= 0,
ratelimiting of duplicate entries is disabled.
.br
If
.I num
= -1, duplicate entries are suppressed indefinitely. See
.B RATELIMITING
section for more details.
.TP
.BI "-H " "size" ", --hashsize=" "size"
Size of ratelimiting hash table used to store recent ethernet/ip address
pairings. Minimum value is 1, maximum value is 65536. Default value is 1 (no
hash table), all pairings are stored in a linked list.

The size of hash table should be increased in active networks with many nodes.
Lager hash table speeds up cache lookups at the cost of increased memory usage.
.TP
.BI "-d, --daemon"
Become a daemon after start. This option implies
.BR "-q" " (" "--quiet" ")"
argument.
.TP
.BI "-h " "hostname" ", --hostname=" "hostname"
Override system hostname.
.TP
.BI "-p " "pidfile" ", --pid=" "pidfile"
Write process id to
.IR "pidfile" "."
Pidfile is deleted when application is terminated.
This option is usually used with 
.BR "-d" " (" "--daemon" ")"
argument. When used with
.BR "-u" " (" "--user" ")"
argument
.I pidfle
will be created before changing active user and in most cases
.B addrwatch
will not be able to delete
.I pidfile
on exit.
.TP
.BI "-P, --no-promisc"
Disable promisc mode on network interfaces.
.TP
.BI "-u " "user" ", --user=" "user"
Suid to 
.I user
after opening network interfaces.
.TP
.BI "-?, --help"
Show command line arguments.
.TP
.BI "--usage"
Give a short usage message.
.TP
.BI "-V, --version"
Print program version.
.SH OUTPUT
In output file or 
.B stdout
each line represents one ethernet/ip address pairing discovery event. Event has
following format
.BR "unix timestamp" ", " "interface" ", " "vlan tag" ", " "mac address" ", "
.B ip address
and
.B packet type
separated by a space.
.PP
Packets without vlan tag is represented by tag value 0.
.PP
Possible packet types are ARP_REQ, ARP_REP, ARP_ACD, ND_ND, ND_NA, ND_DAD.
Ethernet address and IP address are extracted from different parts of the
packet based on packet type.
.IP ARP_REQ 
ARP Request packet. Sender hardware address (from ARP header) and
sender protocol address (from ARP header) is saved.
.IP ARP_REP 
ARP Reply packet. Sender hardware address (from ARP header) and
sender protocol address (from ARP header) is saved.
.IP ARP_ACD 
ARP Address collision detection packet. Sender hardware address
(from ARP header) and target protocol address (from ARP header) is saved.
.IP ND_NS 
Neighbor Solicitation packet.	Source link-layer address (from NS
option) and source address (from IPv6 header) is saved.
.IP ND_NA 
Neighbor Advertisement packet. Target link-layer address (from NA
option) and source address (from IPv6 header) is saved.
.IP ND_DAD 
Duplicate Address Detection packet. Source MAC (from Ethernet
header) and target address (from NS header) is saved.
.PP
Output example:
.nf
.in +4
.B timestamp iface vlan mac ip type
1339405924 eth0 502 e0:ca:94:30:06:8b 2001:db8:200:4202:8946:e6b7:976a:cef3 ND_NA
1339406009 eth0 257 f0:4d:a2:2e:ad:0d 169.254.227.85 ARP_ACD
1339406018 eth0 502 58:1f:aa:d0:92:7d 2001:db8:200:4202:805d:b5ae:8374:436c ND_DAD
1339406029 eth0 502 38:59:f9:3a:de:65 172.16.2.19 ARP_REP
1339406030 eth0 12 00:c0:ee:5a:89:c7 10.1.32.221 ARP_REQ
1339406030 eth0 252 00:1c:c0:79:ab:ee 2001:db8:200:2381::657b ND_NS
.in -4
.fi
.SH RATELIMITING
If used without ratelimiting addrwatch reports etherment/ip pairing every time
it gets usable ARP or IPv6 ND packet. In actively used networks it generates
many duplicate pairings especially for routers and servers.
.PP
Ratelimiting option
.BI "-r " "num"
(
.BI "--ratelimit=" "num"
) suppress output of
.B duplicate
pairings for at least NUM seconds (all non duplicate pairings will be
reported). In other words if addrwatch have discovered
some pairing (mac,ip) it will not report (mac,ip) again unless NUM seconds have
passed.
.PP
There is an exception to this rule to allow tracking ethernet address changes.
If
.B addrwatch
have discovered pairings: (mac1,ip),(mac2,ip),(mac1,ip) within
ratelimit time window it will report all three pairings. By doing so
ratelimiting will not loose any information about ethernet address changes.
.PP
For example if we have a stream of events:
.nf
.in +4
.B "time   ethernet          ip"
01     11:22:33:44:55:66 192.168.0.1
15     11:22:33:44:55:66 192.168.0.1
20     aa:bb:cc:dd:ee:ff 192.168.0.1
25     aa:bb:cc:dd:ee:ff 192.168.0.1
30     11:22:33:44:55:66 192.168.0.1
35     11:22:33:44:55:66 192.168.0.1
40     aa:bb:cc:dd:ee:ff 192.168.0.1
65     aa:bb:cc:dd:ee:ff 192.168.0.1
.in -4
.fi
.PP
With
.B "--ratelimit=100"
we would get:
.nf
.in +4
.B "time   ethernet          ip"
01     11:22:33:44:55:66 192.168.0.1
20     aa:bb:cc:dd:ee:ff 192.168.0.1
30     11:22:33:44:55:66 192.168.0.1
40     aa:bb:cc:dd:ee:ff 192.168.0.1
.in -4
.fi
.PP
Without the exception output would be:
.nf
.in +4
.B "time   ethernet          ip"
01     11:22:33:44:55:66 192.168.0.1
20     aa:bb:cc:dd:ee:ff 192.168.0.1
.in -4
.fi
.PP
And we would loose information that address 192.168.0.1 was used by
11:22:33:44:55:66 between 30-40th seconds.
.PP
To sum up ratelimiting reduces amount of duplicate information without loosing
any ethernet address change events.
.PP
Ratelimit option essentially limits data granularity for IP address usage
duration information (when and for what time period specific IP address was
used). On the other hand without ratelimiting at all you would not get very
precise IP address usage duration information anyways because some hosts might
use IP address without sending ARP or ND packets as often as others
do.
.PP
If
.I num
is set to 0, ratelimiting is disabled and all pairing discovery events
are reported.
.PP
If
.I num
is set to -1, ratelimiting is enabled with infinitely long time window
therefore all duplicate pairings are suppressed indefinitely. In this mode
addrwatch acts almost as arpwatch with the exception that ethernet address
changes are still reported.
.PP
It might look tempting to always use addrwatch with
.B "--ratelimit=-1"
however by
doing so you loose the information about when and for what period of time
specific IP address was used. There will be no difference between temporary
IPv6 addressed which was used once and statically configured permanent
addresses.
.SH NOTES
You can send
.B SIGHUP
signal to
.B addrwatch
to force it reopen output file. It may be useful if used in combination with
.BR logrotate (8).
.SH EXAMPLES
Start monitoring first non loopback network interface. Output will be send to
.BR stdout ":"
.IP
.B addrwatch
.PP
Start monitoring on multiple interfaces:
.IP
.B addrwatch eth0 eth1 eth2
.PP
Start addrwatch as a daemon, write output to file, enable ratelimiting to 1
event per minute:
.IP
.B addrwatch -d -o /var/lib/addrwatch.dat -r 60 eth0
.PP
Start as a daemon, save pid file, send output to syslog, ratelimit to 1 event
per hour, suid to nobody, monitor multiple interfaces:
.IP
.B addrwatch -d -p /var/run/addrwatch.pid -s -r 3600 -u nobody eth0 eth1 eth3
.SH SEE ALSO
.BR logrotate (8)


NEWS
Version 0.7

  * MySQL database output support.

Version 0.6

  * Fixed configure script and source code to compile on FreeBSD.

Version 0.5

  * Fixed ratelimiting with vlan id > 255 bug.

Version 0.4

  * Ratelimiting cache can use hash table to improve speed.
  * Man page.

Version 0.3

  * Added VLAN (802.1Q) support.
  * Sqlite DB schema has new vlan_tag field and is incompatible with previous addrwatch versions.
  * New '-p FILE' argument to save PID to FILE.

Version 0.2

  * IP address blacklisting
  * daemon startup flag
  * malformed packet warnings dumps packet to syslog. With scripts/dump2pcap.pl 
    script can be converted to pcap dump files.


Version 0.1

  * Release of first version with lots of bugs.



src/addrwatch.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <strings.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <argp.h>

#include <sys/stat.h>

#include "addrwatch.h"
#include "parse.h"
#include "check.h"
#include "process.h"
#include "util.h"
#include "mcache.h"
#include "storage.h"
#include "output_flatfile.h"
#include "output_sqlite.h"
#include "output_mysql.h"
#include "output_shm.h"

const char *argp_program_version = PACKAGE_STRING;
const char *argp_program_bug_address = PACKAGE_BUGREPORT;
static char args_doc[] = "[INTERFACE1 INTERFACE2 ...]";
static char doc[] =
"Keep track of ethernet/ip address pairings for IPv4 and IPv6.\
\vIf no interfaces given, then first non loopback interface is used. IP \
address blacklisting opetion '-b' can be used multiple times.";

static struct argp_option options[] = {
	{0, 0, 0, 0, "Options for data output:" },
	{"shm-log-size",'L',"NUM", 0, "Change shared memory log size (default: " "DEFAULT_SHM_LOG_SIZE" ")." },
	{"syslog",    'l', 0,      0, "Output data to syslog (daemon facility)." },
	{"output",    'o', "FILE", 0, "Output data to plain text FILE." },
	{"quiet",     'q', 0,      0, "Suppress any output to stdout and stderr." },
	{"verbose",   'v', 0,      0, "Enable debug messages." },
#if HAVE_LIBMYSQLCLIENT
	{"mysql",     'm', "DB",      OPTION_ARG_OPTIONAL, "Output data to MySQL database DB (default is DB from ~/.my.cnf)." },
	{"mysql-table", 1, "TBL",  0, "Use MySQL table TBL (default: " PACKAGE ")." },
#endif
#if HAVE_LIBSQLITE3
	{"sqlite3",   's', "FILE", 0, "Output data to sqlite3 database FILE." },
	{"sqlite3-table", 2, "TBL",  0, "Use sqlite table TBL (default: " PACKAGE")." },
#endif
	{0, 0, 0, 0, "Options for data filtering:" },
	{"ipv4-only", '4', 0,      0, "Capture only IPv4 packets." },
	{"ipv6-only", '6', 0,      0, "Capture only IPv6 packets." },
	{"blacklist", 'b', "IP",   0, "Ignore pairings with specified IP." },
	{"ratelimit", 'r', "NUM",  0, "Ratelimit duplicate ethernet/ip pairings to 1 every NUM seconds. If NUM = 0, ratelimiting is disabled. If NUM = -1, suppress duplicate entries indefinitely. Default is 0." },
	{"hashsize",  'H', "NUM",  0, "Size of ratelimit hash table. Default is 1 (no hash table)." },
	{0, 0, 0, 0, "Misc options:" },
	{"daemon",    'd', 0,      0, "Start as a daemon." },
	{"pid",       'p', "FILE", 0, "Write process id to FILE." },
	{"no-promisc",'P', 0,      0, "Disable promisc mode on network interfaces." },
	{"user",      'u', "USER", 0, "Suid to USER after opening network interfaces." },
	{"hostname",  'h', "HOSTNAME", 0, "Use HOSTNAME instead of real machine hostname." },
	{ 0 }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static struct argp argp = { options, parse_opt, args_doc, doc };

struct addrwatch_config  cfg;

static const char ip4_filter[] = "arp";
static const char ip6_filter[] = "ip6 and not tcp and not udp and not esp and not ah";
static const char def_filter[] = "ip6 and not tcp and not udp and not esp and not ah or arp";

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{

	switch(key) {
	case '4':
		cfg.v4_flag = 1;
		cfg.v6_flag = 0;
		break;
	case '6':
		cfg.v6_flag = 1;
		cfg.v4_flag = 0;
		break;
	case 'b':
		blacklist_add(arg);
		break;
	case 'd':
		cfg.daemon_flag = 1;
		cfg.quiet = 1;
		break;
	case 'H':
		cfg.hashsize = atoi(arg);
		break;
	case 'L':
		cfg.shm_data.size = atoi(arg);
		if (cfg.shm_data.size < 1)
			cfg.shm_data.size = 1;
		break;
	case 'l':
		cfg.syslog_flag = 1;
		break;
	case 'o':
		cfg.data_file = arg;
		break;
	case 'p':
		cfg.pid_file = arg;
		break;
	case 'P':
		cfg.promisc_flag = 0;
		break;
	case 'q':
		cfg.quiet = 1;
		break;
	case 'r':
		cfg.ratelimit = atoi(arg);
		if (cfg.ratelimit < -1)
			cfg.ratelimit = -1;
		break;
#if HAVE_LIBSQLITE3
	case 's':
		cfg.sqlite_file = arg;
		break;
	case 2:
		cfg.sqlite_table = arg;
		break;
#endif
#if HAVE_LIBMYSQLCLIENT
	case 'm':
		cfg.mysql_flag = 1;
		if (arg)
			cfg.mysql_db = arg;
		break;
	case 1:
		cfg.mysql_table = arg;
		break;
#endif
	case 'u':
		cfg.uname = arg;
		break;
	case 'h':
		cfg.hostname = strdup(arg);
		cfg.hostname_len = strlen(arg) + 1;
		break;
	case 'v':
		cfg.verbose_flag = 1;
		break;
	default:
		return ARGP_ERR_UNKNOWN;
		break;
	}

	return 0;
}

void drop_root(const char *uname)
{
	struct passwd *pw;

	pw = getpwnam(uname);

	if(!pw)
		log_msg(LOG_ERR, "User %s not found", uname);
	
	if (initgroups(uname, pw->pw_gid) != 0
		|| setgid(pw->pw_gid) != 0
		|| setuid(pw->pw_uid) != 0)
		log_msg(LOG_ERR, "Unable to setuid to %s, uid=%d, gid=%d",
			uname, pw->pw_uid, pw->pw_gid);

	log_msg(LOG_DEBUG, "Changed user to %s, uid = %d, gid = %d",
		uname, pw->pw_uid, pw->pw_gid);
}

void pcap_callback(uint8_t *args, const struct pcap_pkthdr *header, const uint8_t *packet)
{
	struct pkt	p;
	int	rc;

	bzero(&p, sizeof(p));

	p.raw_packet = (uint8_t *)packet;

	p.pos = (uint8_t *)packet;
	p.len = header->caplen;

	p.ifc = (struct iface_config *) args;
	p.pcap_header = header;

	rc = parse_packet(&p);

	if (rc < 0)
		return;

	if (p.arp) {
		if (!check_arp(&p))
		process_arp(&p);
	} else if (p.ns) {
		if (!check_ns(&p))
			process_ns(&p);
	} else if (p.na) {
		if (!check_na(&p))
			process_na(&p);
	}

}

#if HAVE_LIBEVENT2
void read_cb(evutil_socket_t fd, short events, void *arg)
#else
void read_cb(int fd, short events, void *arg)
#endif
{
	struct pcap_pkthdr	header;
	const uint8_t	*packet;
	struct iface_config	*ifc;

	ifc = (struct iface_config *) arg;
	packet = pcap_next(ifc->pcap_handle, &header);

	if(packet)
		pcap_callback(arg, &header, packet);
}

void add_iface(char *iface)
{
	char errbuf[PCAP_ERRBUF_SIZE];
	char const *filter;
	struct iface_config *ifc;
	int rc;

	if (cfg.v4_flag)
		filter = ip4_filter;
	else if (cfg.v6_flag)
		filter = ip6_filter;
	else
		filter = def_filter;

	ifc = (struct iface_config *) calloc(1, sizeof(struct iface_config));

	ifc->name = iface;

	ifc->pcap_handle = pcap_open_live(iface, SNAP_LEN, cfg.promisc_flag, 1000, errbuf);
	if (ifc->pcap_handle == NULL) {
		log_msg(LOG_WARNING, "Skipping interface %s, %s", iface, errbuf);
		goto error;
	}

	rc = pcap_datalink(ifc->pcap_handle);
	if (rc != DLT_EN10MB) {
		log_msg(LOG_WARNING, "Skipping interface %s, invalid data link layer %s (%s).", 
			iface,
			pcap_datalink_val_to_name(rc),
			pcap_datalink_val_to_description(rc));
		goto error_pcap;
	}

	rc = pcap_compile(ifc->pcap_handle, &ifc->pcap_filter, filter, 0, 0);
	if (rc == -1) {
		log_msg(LOG_WARNING, "Skipping interface %s, %s",
			iface, pcap_geterr(ifc->pcap_handle));
		goto error_pcap;
	}

	rc = pcap_setfilter(ifc->pcap_handle, &ifc->pcap_filter);
	if (rc == -1) {
		log_msg(LOG_WARNING, "Skipping iface %s, %s",
			iface, pcap_geterr(ifc->pcap_handle));
		goto error_filter;
	}

	rc = pcap_fileno(ifc->pcap_handle);

#if HAVE_LIBEVENT2
	ifc->event = event_new(cfg.eb, rc, EV_READ|EV_PERSIST, read_cb, ifc);
	if(!ifc->event)
		log_msg(LOG_ERR, "%s: event_new(...)", __FUNCTION__);

	event_add(ifc->event, NULL);
#else
	event_set(&ifc->event, rc, EV_READ|EV_PERSIST, read_cb, ifc);
	event_add(&ifc->event, NULL);
#endif

	if (cfg.hashsize < 1 || cfg.hashsize > 65536)
		log_msg(LOG_ERR, "%s: hash size (%d) must be >= 1 and <= 65536", __FUNCTION__, cfg.hashsize);

	if (cfg.ratelimit) {
		ifc->cache = calloc(cfg.hashsize, sizeof(*ifc->cache));
		if (!ifc->cache)
			log_msg(LOG_ERR, "%s: unable to allocate memory for hash cache", __FUNCTION__);
	}

	ifc->next = cfg.interfaces;
	cfg.interfaces = ifc;

	log_msg(LOG_DEBUG, "Opened interface %s (%s)",
		iface,
		pcap_datalink_val_to_description(pcap_datalink(ifc->pcap_handle)));

	return;

error_filter:
	pcap_freecode(&ifc->pcap_filter);
error_pcap:
	pcap_close(ifc->pcap_handle);
error:
	free(ifc);
}

struct iface_config *del_iface(struct iface_config *ifc)
{
	struct iface_config	*next;
	int	i;

	next = ifc->next;

#if HAVE_LIBEVENT2
	event_free(ifc->event);
#endif
	pcap_freecode(&ifc->pcap_filter);
	pcap_close(ifc->pcap_handle);

	log_msg(LOG_DEBUG, "Closed interface %s", ifc->name);

	if (ifc->cache) {
		for (i = 0; i < cfg.hashsize; i++)
			if (*(ifc->cache + i))
				cache_prune(*(ifc->cache+i), ifc->cache+i);
		free(ifc->cache);
	}

	free(ifc);
	
	return next;

}

#if HAVE_LIBEVENT2
void reload_cb(evutil_socket_t fd, short events, void *arg)
#else
void reload_cb(int fd, short events, void *arg)
#endif
{
	log_msg(LOG_DEBUG, "Received signal (%d), %s", fd, sys_siglist[fd]);
	log_msg(LOG_DEBUG, "Reopening output files");

	output_flatfile_reload();
	output_sqlite_reload();
	output_mysql_reload();
	output_shm_reload();
}

#if HAVE_LIBEVENT2
void stop_cb(evutil_socket_t fd, short events, void *arg)
#else
void stop_cb(int fd, short events, void *arg)
#endif
{
	log_msg(LOG_DEBUG, "Received signal (%d), %s", fd, sys_siglist[fd]);
#if HAVE_LIBEVENT2
	event_base_loopbreak(cfg.eb);
#else
	event_loopbreak();
#endif
}



void libevent_init()
{
#if HAVE_LIBEVENT2
	cfg.eb = event_base_new();

	if (!cfg.eb)
		log_msg(LOG_ERR, "%s: event_base_new() failed", __FUNCTION__);
#else
	event_init();
#endif

#if HAVE_LIBEVENT2
	cfg.sigint_ev = event_new(cfg.eb, SIGINT, EV_SIGNAL|EV_PERSIST, stop_cb, NULL);
	event_add(cfg.sigint_ev, NULL);
#else
	event_set(&cfg.sigint_ev, SIGINT, EV_SIGNAL|EV_PERSIST, stop_cb, NULL);
	event_add(&cfg.sigint_ev, NULL);
#endif

#if HAVE_LIBEVENT2
	cfg.sigterm_ev = event_new(cfg.eb, SIGTERM, EV_SIGNAL|EV_PERSIST, stop_cb, NULL);
	event_add(cfg.sigterm_ev, NULL);
#else
	event_set(&cfg.sigterm_ev, SIGTERM, EV_SIGNAL|EV_PERSIST, stop_cb, NULL);
	event_add(&cfg.sigterm_ev, NULL);
#endif

#if HAVE_LIBEVENT2
	cfg.sighup_ev = event_new(cfg.eb, SIGHUP, EV_SIGNAL|EV_PERSIST, reload_cb, NULL);
	event_add(cfg.sighup_ev, NULL);
#else
	event_set(&cfg.sighup_ev, SIGHUP, EV_SIGNAL|EV_PERSIST, reload_cb, NULL);
	event_add(&cfg.sighup_ev, NULL);
#endif
}

void libevent_close()
{
#if HAVE_LIBEVENT2
	event_free(cfg.sigint_ev);
	event_free(cfg.sigterm_ev);
	event_free(cfg.sighup_ev);

	event_base_free(cfg.eb);
#endif

}

void save_pid()
{
	FILE	*f;

	if (!cfg.pid_file)
		return;

	f = fopen(cfg.pid_file, "w");
	if(!f)
	{
		log_msg(LOG_ERR, "Unable to open pid file %s", cfg.pid_file);
		return;
	}
	fprintf(f, "%d\n", getpid());
	fclose(f);
}

void del_pid()
{
	if (!cfg.pid_file)
		return;

	unlink(cfg.pid_file);
}

void daemonize()
{
	pid_t pid, sid;

	if (cfg.daemon_flag) {
		pid = fork();

		if (pid < 0)
			log_msg(LOG_ERR, "Error forking new process");

		if (pid > 0)
			exit(EXIT_SUCCESS);

		umask(0);
		sid = setsid();
		if (sid < 0)
			log_msg(LOG_ERR, "Error starting new session");
	}
	
	fclose(stdin);
	if (cfg.quiet) {
		fclose(stdout);
		fclose(stderr);
	}
}

int main(int argc, char *argv[])
{
	char errbuf[PCAP_ERRBUF_SIZE];
	char *dev;
	struct iface_config *ifc;
	int optind;
	int i;


	bzero(&cfg, sizeof(cfg));

	/* Default configuration */
//	cfg.ratelimit = 0;
	cfg.hashsize = 1;
//	cfg.quiet = 0;
	cfg.promisc_flag = 1;
//	cfg.ratelimit = 0;
//	cfg.sqlite_file = NULL;
//	cfg.uname = NULL;
	cfg.shm_data.size = DEFAULT_SHM_LOG_SIZE;
#if HAVE_LIBSQLITE3
	cfg.sqlite_table = PACKAGE;
#endif
#if HAVE_LIBMYSQLCLIENT
//	cfg.mysql_db = NULL;
	cfg.mysql_table = PACKAGE;
#endif

	argp_parse(&argp, argc, argv, 0, &optind, 0);

	if (!cfg.hostname) {
		cfg.hostname_len = sysconf(_SC_HOST_NAME_MAX);
		cfg.hostname = (char *)calloc(cfg.hostname_len, sizeof(char));
		gethostname(cfg.hostname, cfg.hostname_len);
	}

	daemonize();
	save_pid();

	log_open();
	libevent_init();


	if (cfg.ratelimit > 0)
		log_msg(LOG_DEBUG, "Ratelimiting duplicate entries to 1 per %d seconds", cfg.ratelimit);
	else if (cfg.ratelimit == -1)
		log_msg(LOG_DEBUG, "Duplicate entries supressed indefinitely");
	else
		log_msg(LOG_DEBUG, "Duplicate entries ratelimiting disabled");

	if (cfg.promisc_flag)
		log_msg(LOG_DEBUG, "PROMISC mode enabled");
	else
		log_msg(LOG_DEBUG, "PROMISC mode disabled");

	if (argc > optind) {
		for (i = optind; i < argc; i++)
			add_iface(argv[i]);
	} else {
		dev = pcap_lookupdev(errbuf);
		if (dev != NULL)
			add_iface(dev);
	}

	if (!cfg.interfaces)
		log_msg(LOG_ERR, "No suitable interfaces found!");

	if (cfg.uname)
		drop_root(cfg.uname);

	output_flatfile_init();
	output_sqlite_init();
	output_mysql_init();
	output_shm_init();

	/* main loop */
#if HAVE_LIBEVENT2
	event_base_dispatch(cfg.eb);
#else
	event_dispatch();
#endif

	output_shm_close();
	output_mysql_close();
	output_sqlite_close();
	output_flatfile_close();

	for (ifc = cfg.interfaces; ifc != NULL; ifc = del_iface(ifc));


	libevent_close();
	log_close();

	del_pid();
	blacklist_free();

	free(cfg.hostname);

	return 0;
}
